Feature/shares management forms controller

AlvaroGRuiz opened the pull request:
Feature/shares management forms controller

implements:
* OP#417
* OP#1707
* OP#1579
* OP#1816
* OP#1837
* OP#1967

AlvaroGRuiz committed:
[IMP] ✨ New onchange for product on become_cooperator

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Change label and help for field display_on_website at product.template form

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[I18N] 🌐 Add translate for field display_on_website

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Add new fields url_individual and url_company for product.template

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Add selector of product for website BASE

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ New field boolean for Product selector in BASE form

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ New fields on company and products. External id for company and products. Controller become_cooperator and share_subscription load company for external_id. Selector of product show for configuration.

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Changes on product view

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Invitations share setting on form company and res_confi_settings

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

daniquilez committed:
[WIP] 🚧 Draft base structure for generic subscription controller

AlvaroGRuiz committed:
[IMP] ✨ Estructure of controller and views to form subcription

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

daniquilez committed:
[WIP] 🚧 New subscription controller base structure

 Co-authored-by: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[WIP] 🚧 Controller and templates views for subcription cooperator request

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

daniquilez committed:
[WIP] 🚧 Next steps from subscription controller. Move class variables to values

AlvaroGRuiz committed:
[IMP] ✨ Data of controllers for subcriptions forms and views.

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Fix names of templates

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

juan-pe commented on energy_communities_cooperator/controllers/website_share_subscription_controller.py:
```diff
@@ -0,0 +1,233 @@
+# http://odoo-ce.local:8069/es/subscription/company_invited/356a192b7913b04c54574d18c28d46e6395428ab/ae7329c979b3cd96086c22cca6217764ab3e50ec
+# http://odoo-ce.local:8069/es/subscription/company_voluntary/356a192b7913b04c54574d18c28d46e6395428ab/ae7329c979b3cd96086c22cca6217764ab3e50ec
+# http://odoo-ce.local:8069/es/subscription/company_member/356a192b7913b04c54574d18c28d46e6395428ab/ae7329c979b3cd96086c22cca6217764ab3e50ec
+class WebsiteShareSubscriptionController(http.Controller):
```
@daniquilez @AlvaroGRuiz 
Some considerations about this controller :)

I think is a better idea, write in the same function `GET` and `POST`. We don't duplicate code and the benefits of this are, fewer errors, less cognitive load, easier maintenance. 
In fact, we have an error in this code, in `GET` function we validate the request, when this should be done in `POST`. I think that `subscription_page` and `subscription_submission` are interchanged 😬 😬 

I propose something like:
```python
def share_subscription_form(self, **kwargs):
    values = self._get_base_values(kwargs)
    if request.method == "POST":
        try:
            self._validate_request(values)
        except ValidationFormError as e:
            self._get_page_values(values)
            self._get_form_fields(values)
            self._get_custom_options(values)
            self._get_custom_description(values)
            return self._render_page(values)
        else:
            self.process_form(values)
            return self._render_successful_page()
    return self._render_page(values)
```

The proposal is to structure the code in blocks:
* discriminate the method
* if post:
  * validate the request
  * process the form
  * return to form if errors or to a successful page if every thing is ok
* if get:
  * show form

AlvaroGRuiz replied:
i propose:
@http.route(
        [
            "/subscription/<string:mode>/<string:company_ext_id>",
            "/subscription/<string:mode>/<string:company_ext_id>/<string:product_ext_id>",
        ],
        type="http",
        auth="public",
        website=True,
        methods=["GET", "POST"],
    )
    def share_subscription_form(self, **kwargs):
        values = self._get_base_values(kwargs)
        if request.httprequest.method == "GET":
            self._validate_request(values)
            self._get_page_values(values)
            self._get_form_fields(values)
            self._get_custom_options(values)
            self._get_custom_description(values)
            return self._render_page(values)
        self._process_form(values)
        return self._render_page(values)
        
   I dont know _process_form on GET have to do, I think is on method POST.
   Request.method dont exist, it is request.httprequest.method.
   ValidationFormError dont exit. If i use try, form dont work.

juan-pe commented on energy_communities_cooperator/controllers/website_share_subscription_controller.py:
```diff
@@ -0,0 +1,233 @@
+            currency_symbol=values["currency_symbol"],
+        )
+
+    def _get_form_fields_values(self, data, path):
```
I don't really understand what is the meaning of this method... also, without the context, it could be nice to return the same type of value. Looks like if every thing is ok, this function will return a dictionary, so when there was same error, return a dictionary too.

AlvaroGRuiz replied:
This function allows you to retrieve values from objects within our dictionary of values. For example, I want a field to have the value of company.name. Since we have the object company in the dictionary of values with the path company.name, it is able to find the value. If it is not an object, it returns the assigned value, for example “hello,” True, 1, 0.75, etc.

juan-pe commented on energy_communities_cooperator/controllers/website_share_subscription_controller.py:
```diff
@@ -0,0 +1,233 @@
+            current = current[key]
+        return current
+
+    def _get_form_fields(self, values):
```
I should rename the following methods (`_get_form_fields`, `_get_custom_options`, `_get_custom_description`, ` _get_page_values`)  to `_update`, instead of `_get`, because they aren't returning anything, but they are updating `values` dictionary

AlvaroGRuiz replied:
oki

juan-pe commented on energy_communities_cooperator/controllers/website_subscription_main.py:
```diff
@@ -32,18 +32,76 @@ def display_become_cooperator_page(self, **kwargs):
-        if ("odoo_company_id" in kwargs) and (
-            not target_odoo_company_id
-            or not request.env["res.company"]
+        if kwargs.get("external_company_id", False):
```
We repeat this code several times, I think is a good idea to write it in a common function ;)

AlvaroGRuiz replied:
I dont see it

AlvaroGRuiz replied:
we dont use it

juan-pe commented on energy_communities_cooperator/controllers/website_subscription_main.py:
```diff
@@ -32,18 +32,76 @@ def display_become_cooperator_page(self, **kwargs):
+            except:
+                pass
+
+        if (
```
It's hard to understand the meaning of this condition... I propose to take out of the `if` and write a variable that express the meaning of it.

Also, this should be a [400](https://http.cat/status/400) response, it doesn't like that this is an internal server error 😆

AlvaroGRuiz replied:
I dont see it

AlvaroGRuiz replied:
we dont use it

juan-pe commented on energy_communities_cooperator/controllers/website_subscription_main.py:
```diff
@@ -32,18 +32,76 @@ def display_become_cooperator_page(self, **kwargs):
+            except:
+                pass
+
+        if ("product_external_id" in kwargs) and (
```
Same as the previous comment

AlvaroGRuiz replied:
we dont use it

juan-pe commented on energy_communities_cooperator/controllers/website_subscription_main.py:
```diff
@@ -58,18 +116,75 @@ def display_become_company_cooperator_page(self, **kwargs):
-        if ("odoo_company_id" in kwargs) and (
-            not target_odoo_company_id
-            or not request.env["res.company"]
+        if kwargs.get("external_company_id", False):
```
Same as the previous comment. Looks like we can write all this code in a validation method.

AlvaroGRuiz replied:
we dont use it

juan-pe commented on energy_communities_cooperator/controllers/website_subscription_main.py:
```diff
@@ -84,14 +199,40 @@ def share_subscription(self, **kwargs):
-            or not request.env["res.company"]
-            .sudo()
-            .search([("id", "=", target_odoo_company_id)])
+        if kwargs.get("external_company_id", False):
```
Same as the previous comment

AlvaroGRuiz replied:
we dont use it

juan-pe commented on energy_communities_cooperator/config.py:
```diff
@@ -1,4 +1,193 @@
+    "voluntary": COOP_VOLUNTARY_SHARE_PRODUCT_CATEG_REF,
+    "company_voluntary": COOP_VOLUNTARY_SHARE_PRODUCT_CATEG_REF,
+}
+MAPPING__SUBSCRIPTION_MODE__DEFAULT_PAGE_TITLE = {
```
I don't understand why should this code be in two strings instead of one.
`_("Become Cooperator of {company_name} ") + _("member_eu_ES")`

AlvaroGRuiz replied:
We have translate with term member_eu_ES. If we merge it, we need a new translate.

juan-pe commented on energy_communities_cooperator/controllers/website_subscription_main.py:
```diff
@@ -114,6 +258,14 @@ def fill_values(self, values, is_company, logged, load_from_user=False):
             company = request.website.company_id
         products = self.get_products_share(is_company)
 
+        product = False
```
The following lines look like "dead code". 
If there is a `target_product_externnal_id` we set `product` variable, but later in the code, we overwrite `product` during the `for` sentence in line 285...

AlvaroGRuiz replied:
we dont use it

juan-pe commented on energy_communities_cooperator/controllers/website_subscription_voluntary_share.py:
```diff
@@ -37,14 +37,40 @@ def display_voluntary_share_page(self, **kwargs):
-            or not request.env["res.company"]
-            .sudo()
-            .search([("id", "=", target_odoo_company_id)])
+        if kwargs.get("external_company_id", False):
```
Same comment has the previous controller about this validation.

AlvaroGRuiz replied:
I dont see it

AlvaroGRuiz replied:
we dont use it

juan-pe commented on energy_communities_cooperator/models/product.py:
```diff
@@ -1,18 +1,63 @@
 class ProductTemplate(models.Model):
     _inherit = "product.template"
 
+    def _compute_url(self):
```
It could be less error-prone to use [urllib](https://docs.python.org/3/library/urllib.html) from the python stdlib to generate urls.

To build the query string part, use the function [urlencode](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urlencode)
To join an url, use [urljoin](https://docs.python.org/3/library/urllib.parse.html#urllib.parse.urljoin) function

juan-pe commented on energy_communities_cooperator/models/res_company.py:
```diff
@@ -6,6 +8,19 @@ class ResCompany(models.Model):
+            ).hexdigest()
+
+    def _compute_share_urls(self):
+        for record in self:
```
same as the previous comment.
Also extract `self.env['ir.config_parameter'].sudo().get_param('web.base.url')` to a variable and call only once.

AlvaroGRuiz replied:
oki

juan-pe commented on energy_communities_cooperator/static/src/js/cooperator.js:
```diff
@@ -1,8 +1,38 @@
+          .jsonRpc("/subscription/get_share_product", "call", {
+            share_product_id: share_product_id,
+          })
+          .then(function (data) {
```
what happens if the call fails?

AlvaroGRuiz replied:
We dont use it now

AlvaroGRuiz committed:
[IMP] ✨ Implement method payment on forms

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

juan-pe commented on energy_communities_cooperator/views/website_template_subscription_base.xml:
```diff
@@ -0,0 +1,69 @@
+
+  <template id="template_subscription_page" name="template_subscription_page">
+    <t t-call="website.layout">
+      <script
```
As a good practice, javascript scripts and libraries should go at the end of the html, just when the body ends.
css files, instead should be at the beginning, in the `head` section

AlvaroGRuiz replied:
ok

juan-pe reviewed (changes requested):
Wow! what an incredible work! 🎉 

There is a lot of work here and most of them looks very nice!

I wrote some comments that would be nice to have a conversation about them. 

Thanks a lot!! 🔝

AlvaroGRuiz committed:
[IMP] ✨ Delete mode company voluntary, we dont need it

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Implement prosose by Juan Pedro Sánchez

Author: Alvaro Garcia <alvaro.garcia@somit.coop>
Author: Juan Pedro Sánchez <juanpe.sanchez@somenergia.coop>

AlvaroGRuiz committed:
[IMP] ✨ Logisc specific forms

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Add logic for product price == 0

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Add language at forms

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

daniquilez committed:
[WIP] 🚧 Controller validation and submission adjusted to pydantic and components

 Co-authored-by: Alvaro Garcia <alvaro.garcia@somit.coop>
 Co-authored-by: Juanpe Sánchez  <juanpe.sanchez@somenergia.coop>

AlvaroGRuiz committed:
[WIP] 🚧 Fix names

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[WIP] 🚧 Validation request

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Fix errors on pydantic class

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Fix api depends on product and company

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Error search product

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

daniquilez committed:
[WIP] 🚧 Next steps on controller structure

AlvaroGRuiz committed:
[FIX] 🐛 Create values is better and templates support ctx

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

daniquilez committed:
[WIP] 🚧 More code style changes on controller

daniquilez committed:
[TEST] ✅ All wrong url scenarios

 Co-authored-by: Alvaro Garcia <alvaro.garcia@somit.coop>

daniquilez committed:
[TEST] ✅ Introducing consistency validation

 Co-authored-by: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ News validations, logic to select products and new data demo

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

daniquilez committed:
[TEST] ✅ Full controller context validation: NotFound, Consistency, Availability

daniquilez committed:
[WIP] 🚧 Form must be loaded only trough config file

AlvaroGRuiz committed:
[IMP] ✨ Now the field form is dinamic, you can implement by confi values

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ All code of controller are comment, better implement of extra options in input select

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[WIP] 🚧 component for subcription request

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

juan-pe committed:
[REF] ♻️  call subscription request util component like the others in the plantform

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>
Co-authored-by: Dani Quilez  <dani.quilez@somenergia.coop>

juan-pe committed:
[REF] ♻️  write controller with structure request validation - process request - return response

We want to structure our controllers in the way:
1.- request validation: can our controler understand your request?
2.- process request: If we understand what are you request to us, we
  will process the request or show you some errors it the request isn't
  correct
3.- return response: depending if we can process or not your request, we
  will return a sucesfull response or an error one

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>
Co-authored-by: Dani Quilez  <dani.quilez@somenergia.coop>

daniquilez committed:
[WIP] 🚧 _process_form_base_structure

daniquilez committed:
[WIP] 🚧 Add SubscriptionRequestCreationParams Pydantic validation

AlvaroGRuiz committed:
[WIP] 🚧 Process form to the end

Author: Alvaro Garcia <alvaro.garcia@somit.coop>
Author: Dani Quilez <dani.quilez@somenergia.coop>

AlvaroGRuiz committed:
[WIP] 🚧 Process form to the end (Fixed comented part)

Author: Alvaro Garcia <alvaro.garcia@somit.coop>
Author: Dani Quilez <dani.quilez@somenergia.coop>

daniquilez committed:
[IMP] ✨ Better error handling on form controllers

 Co-authored-by: Alvaro Garcia <alvaro.garcia@somit.coop>

daniquilez committed:
[IMP] ✨ Remove please select on selection options and pass htt code to render page

 Co-authored-by: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Company member form view

Author: Alvaro Garcia <alvaro.garcia@somit.coop>
Author: Dani Quilez <dani.quilez@somenergia.coop>

AlvaroGRuiz committed:
[IMP] ✨ Company member form backend

Author: Alvaro Garcia <alvaro.garcia@somit.coop>
Author: Dani Quilez <dani.quilez@somenergia.coop>

AlvaroGRuiz committed:
[WIP] 🚧 Voluntary form backend and frontend

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

daniquilez committed:
[IMP] ✨ Apply consistency validation when subscription_Request might change partner data

daniquilez committed:
[WIP] 🚧 Testing voluntary shares submissions

daniquilez committed:
[IMP] ✨ Fix Company subcription params optional params

AlvaroGRuiz committed:
[IMP] ✨ Implement test_website_form_voluntary_submission_ok

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[REF] ♻️  We dont need this code

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Change dependency of python lib email_validator==2.2.0 to email_validator==2.3.0

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 URLs on product and category product form view and tests of energy_communities_cooperator

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Fix api depends on urls field of product form view

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

juan-pe committed:
[WIP] 🚧 energy_comunities_cooperator: test proposal for validate__member_subscription_request

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>
Co-authored-by: Dani Quilez  <dani.quilez@somenergia.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Fix class TestSubscriptionRequest

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Validations from subcription request component

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ New test on energy communities cooperator

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Implement config by company(privacy_policy, internal_rules, financial_risk, generic_rules, voluntary_share_form_header_text and cooperator_share_form_header_text) and by product (html_specific_products)

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Implement limit of config complany for subscrition (subscription_maximum_amount)

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Implement default_lang_id on url for forms subscriptions

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ New field subscription_mode on subscription.request. New filters on subscription.request

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Subcrition request write gender on res partner

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Fix form controllers

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Implement new logic for cooperator

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Refactor customs form and options

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Refactor buttons for forms views

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Field gender on form view for res_partner

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Fix buttons for URLS forms

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Fix urls

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Fix url voluntary company

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Implement effective_member == True if set inveted category, show membership_type on view form and cooperator_register_member set 0 if is invited

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ if you are invite, you dont are member

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Field data_xml_id for product category now is readonly and invisible. Invited category now have this field filled.

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Delete field assume_already_member.energy_communities.product_category_share_recurring_fee_pack category have ProductWebsite=False and Product Qty Must Be Read Only=True. Fields product_website and product_qty_must_be_read_only are for company.

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ New field email_corfirm and validate on javascript. Fields disable now are readonly(disable flield dont submit on form). Field date now is type=date. Polity Privacity are optional field.

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Fix load field or not for settings

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 H3 now can be invisible

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Change on view products. Change hedear custom text for voluntary forms.

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ Now the category have a header custom for forms

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 On load category need sudo()

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ New subcription mode member_associations and tests for this

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ New text for title and 1º text of forms

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[I18N] 🌐 Translate for mofule energy_communitites_cooperator

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Fix translate on form controller views. Fix load category with company. Copy dictionary static.

Author: Alvaro Garcia <alvaro.garcia@somit.coop>
Author: Juan Pedro Sánchez <juanpe.sanchez@somenergia.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Fix translate

Author: Alvaro Garcia <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[I18N] 🌐 Translate headlines of controller forms

Co-authored-by: Juan Pedro Sánchez <juanpe.sanchez@somenergia.coop>

AlvaroGRuiz committed:
[I18N] 🌐 Translate labels

daniquilez committed:
On card 1579 point 2 Identify where to setup requested configuration

daniquilez committed:
Create invited product on multi company easy creation execution and properly setup category

AlvaroGRuiz committed:
[I18N] 🌐 New translate

AlvaroGRuiz committed:
[I18N] 🌐 Translate in module energy_communitites_cooperator

AlvaroGRuiz committed:
[I18N] 🌐 Translate of modules energy_communities,energy_communities_api,energy_communities_cooperator,energy_communities_crm and energy_communities_service_invoicing

AlvaroGRuiz committed:
[FIX] 🐛 Fix to hide selector products on website form

AlvaroGRuiz committed:
[FIX] 🐛 Fix load text custom for specifict forms

AlvaroGRuiz committed:
[FIX] 🐛 Delete texts member_eu_ES

AlvaroGRuiz committed:
[FIX] 🐛 Fix default on product category for product_website and product_qty_must_be_read_only

AlvaroGRuiz committed:
[IMP] ✨ Implement script post-migration for set feault values for product.category on fields product_website and product_qty_must_be_read_only

AlvaroGRuiz committed:
[IMP] ✨ Implement news titles on website forms

AlvaroGRuiz committed:
[IMP] ✨ News redirect for old urls to news urls website forms

AlvaroGRuiz committed:
[FIX] 🐛 Redirect URLs

AlvaroGRuiz committed:
[FIX] 🐛 Fix attribute readonly on templates

AlvaroGRuiz committed:
[FIX] 🐛 Fix something for frontend (Labes, new labels, product_website and product_qty_must_be_read_only)

AlvaroGRuiz committed:
[IMP] ✨ Implement new form company_member_associations

AlvaroGRuiz committed:
[IMP] ✨ New redirect for company_member_associations forms

AlvaroGRuiz committed:
[REF] ♻️  Refactor redirect URLs for cooperator websites

AlvaroGRuiz committed:
[I18N] 🌐 Apply new translations

AlvaroGRuiz committed:
[IMP] ✨ New query to check invited member on cooperators

Co-authored-by: Juan Pedro Sánchez <juanpe.sanchez@somenergia.coop>

AlvaroGRuiz committed:
[WIP] 🚧 Migration old invited and change test on wizard for new invites

AlvaroGRuiz committed:
[WIP] 🚧 Migration old partner invited for cooperator

juan-pe committed:
[WIP] 🚧 energy_communities_cooperator: working in tests for the migration

./odoo-bin -c config/odoo.conf -p 8079 -d odoo-1 -u energy_communities_cooperator --test-enable --test-tags=/upgrade.test_prepare --upgrade-path=../upgrade-util/src --log-level=info --logfile=/dev/stdout --stop-after-init

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator,energy_communities_service_invoicing: fix component usages

fix in migration energy_communities_service_invoicing components loading, and also added usage in ProductUtils component

AlvaroGRuiz committed:
[IMP] ✨ Implement text creation new inscriptions of type invited

AlvaroGRuiz committed:
[FIX] 🐛 Fix version migrate script

AlvaroGRuiz committed:
[IMP] ✨ New version for migration script on energy_community_cooperator

AlvaroGRuiz committed:
[WIP] 🚧 Session 2026-02-12

AlvaroGRuiz committed:
[FIX] 🐛 Set variable on categories for new company. New fields on socias views.

AlvaroGRuiz committed:
[FIX] 🐛 Fix migration script

AlvaroGRuiz committed:
[I18N] 🌐 Fix translations

AlvaroGRuiz committed:
[I18N] 🌐 Fix translate

AlvaroGRuiz committed:
[FIX] 🐛 Fix translations

AlvaroGRuiz committed:
[FIX] 🐛 Text on form invited

AlvaroGRuiz committed:
[I18N] 🌐 Fix translations

juan-pe committed:
[WIP] 🚧 energy_comunities_cooperator: validation wip

type <description>

juan-pe committed:
[WIP] 🚧 energy_communities_cooperator: WIP validations

wip manage validations and exceptions in corner cases

Commit:
Translated using Weblate (Spanish)

Currently translated at 75.7% (312 of 412 strings)

Translation: odoo/odoo-ce
Translate-URL: http://localhost:8080/projects/odoo/odoo-ce/es/

Commit:
Translated using Weblate (Catalan)

Currently translated at 99.0% (408 of 412 strings)

Translation: odoo/odoo-ce
Translate-URL: http://localhost:8080/projects/odoo/odoo-ce/ca/

Commit:
Translated using Weblate (Basque)

Currently translated at 81.3% (335 of 412 strings)

Translation: odoo/odoo-ce
Translate-URL: http://localhost:8080/projects/odoo/odoo-ce/eu/

AlvaroGRuiz committed:
[I18N] 🌐 New translation and add .pot

Commit:
[I18N][Catalan] 🌐 energy_communities_cooperator: Translated using Weblate

Commit:
[I18N][Basque] 🌐 energy_communities_cooperator: Translated using Weblate

Commit:
[I18N][Spanish] 🌐 energy_communities_cooperator: Translated using Weblate

AlvaroGRuiz committed:
[I18N] 🌐 New .pot for module energy_communitites cooperator

Commit:
Translated using Weblate (Spanish)

Currently translated at 53.6% (141 of 263 strings)

Translation: odoo/energy_communities_service_invoicing
Translate-URL: http://localhost:8080/projects/odoo/energy_communities_service_invoicing/es/

Commit:
Translated using Weblate (Catalan)

Currently translated at 44.1% (116 of 263 strings)

Translation: odoo/energy_communities_service_invoicing
Translate-URL: http://localhost:8080/projects/odoo/energy_communities_service_invoicing/ca/

Commit:
Translated using Weblate (Basque)

Currently translated at 46.0% (121 of 263 strings)

Translation: odoo/energy_communities_service_invoicing
Translate-URL: http://localhost:8080/projects/odoo/energy_communities_service_invoicing/eu/

Commit:
Translated using Weblate (Basque)

Currently translated at 100.0% (129 of 129 strings)

Translation: odoo/energy_communities_cooperator extra
Translate-URL: http://localhost:8080/projects/odoo/energy_communities_cooperator-extra/eu/

Commit:
[I18N] [Spanish] 🌐 Translation for module: odoo/energy_communities

Commit:
Translated using Weblate (Catalan)

Currently translated at 97.0% (291 of 300 strings)

Translation: odoo/energy_communities_cooperator
Translate-URL: http://localhost:8080/projects/odoo/energy_communities_cooperator/ca/

Commit:
Translated using Weblate (Spanish)

Currently translated at 95.9% (287 of 299 strings)

Translation: odoo/energy_communities_cooperator
Translate-URL: http://localhost:8080/projects/odoo/energy_communities_cooperator/es/

Commit:
Translated using Weblate (Basque)

Currently translated at 97.6% (287 of 294 strings)

Translation: odoo/energy_communities_cooperator
Translate-URL: http://localhost:8080/projects/odoo/energy_communities_cooperator/eu/

AlvaroGRuiz committed:
[I18N] 🌐 New term for translation on i18n extra

AlvaroGRuiz committed:
[I18N] 🌐 Translate new term of cooperator on energy_community_cooperator i18n_extra

juan-pe committed:
[REL] 🏗️  update migration version to version 16.0.0.7.0

juan-pe committed:
[TEST] ✅ energy_communities_cooperator: fixed tests

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: fix existing partner validation for voluntary shares

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: iban is not required when payment method is transfer

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>

 # Please enter the commit message for your changes. Lines starting

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: fix default values in website form

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>

juan-pe committed:
[FIX] 🐛 energy_communities: correct succesfull page

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: save check coditions payment

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>

juan-pe committed:
[#1579][REF] ♻️  energy_communities_cooperator: refactor SubscriptionRequestCreationParams creation in order to use this class to validate subscription requests created from odoo

AlvaroGRuiz committed:
[FIX] 🐛 Move fields effective_date,cooperator_type and cooperator

AlvaroGRuiz committed:
[FIX] 🐛 Delete 2 filters on subscription.request in tree view

AlvaroGRuiz committed:
[FIX] 🐛 Fix SubcritionRequestParam and cooperator_number_effective

AlvaroGRuiz committed:
[FIX] 🐛 Fix send email at account_move to paid

AlvaroGRuiz committed:
[FIX] 🐛 number of cooperator and coop_candidate

juan-pe committed:
[REF] ♻️  energy_communities_cooperator: move interelational fields validations to schema;

added invited validation to component

juan-pe committed:
[IMP] ✨ energy_communities_cooperator: add invited validation in subscription request validation

juan-pe committed:
[IMP] ✨ energy_communities_cooperator: subscription request params from odoo forms are now also validated with SubscriptionRequestCreationParams schema

juan-pe committed:
[REF] ♻️  energy_communities_cooperator: extract validation phase to a specific function in subscruption request creation

juan-pe committed:
[REF] ♻️  energy_communities_cooperator: get product category from share product

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>

juan-pe committed:
[FIX] 🐛 energy_comunities_cooperator: share_product_id has to be product.template, but we recive a product.product model on odoo form creation

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>

juan-pe committed:
[TEST] ✅ & 🚧 energy_communities_cooperator: wip subscruption request creation test

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Now you can see effective_invited on view socias and page share if you are effective_invited

AlvaroGRuiz committed:
[I18N] 🌐 Fix translate Convidades a participar

AlvaroGRuiz committed:
[FIX] 🐛 Fix assign_cooperator_register_number function

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: add voluntary shares forgotten attributes

also add forgotten FormValidationError import

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: fix validation message when not loc

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: generic, internal and financial check rules have to be applied to all forms

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: import errors + force is cooperator for voluntary shares

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: do not force already cooperator. fine grained existing vat for cooperator shares

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: fix string representation for custom exceptions

also validationError typo....  :'D

juan-pe committed:
[REF] ♻️  energy_communities_cooperator: unify error management for pydantic exceptions

juan-pe committed:
[CI/CD] 🔧 add editorconfig file

juan-pe committed:
[STYLE] 🎨 energy_communities_cooperator: change error message when partner not foud

Co-authored-by: Xavi Bonet  <xavi.bonet@somenergia.coop>

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: subscriptionrequest component has to work under company context

Co-authored-by: Xavi Bonet  <xavi.bonet@somenergia.coop>

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: _validate_vals should "ear" for FormValidationError exceptions

also added default value to None for company in subscription_request_utils

Co-authored-by: Xavi Bonet  <xavi.bonet@somenergia.coop>

juan-pe committed:
[IMP] ✨ energy_communities_cooperator: for voluntary shares, fullfill partner values in subscription requests

Co-authored-by: Xavi Bonet  <xavi.bonet@somenergia.coop>

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: fix conditions in validation for cooperative shares

Co-authored-by: Xavi Bonet  <xavi.bonet@somenergia.coop>

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: invited company is a WebsiteShareSubscriptionSubmissionCompanyMember form

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: invited company is a WebsiteShareSubscriptionSubmissionCompanyMember form

juan-pe committed:
[IMP] ✨ energy_communities_Cooperator: fix effective_date and cooperator_type visibilty

Co-authored-by: Dani Quilez  <dani.quilez@somenergia.coop>

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: fix filter for mandatory shares

Co-authored-by: Xavi Bonet  <xavi.bonet@somenergia.coop>

juan-pe committed:
[IMP] ✨ energy_communities_cooperator: implement transition of invited to member memberships

Co-authored-by: Xavi Bonet  <xavi.bonet@somenergia.coop>

juan-pe committed:
[IMP] ✨ energy_communities_cooperator: implement transition of invited to member memberships

Co-authored-by: Xavi Bonet  <xavi.bonet@somenergia.coop>

XavierBonetViura committed:
[IMP] ✨ Wizard & button in Product Category form to modify cooperative generic form layout settings

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: fix order record definition in product_category_views

Co-authored-by: Xavi Bonet  <xavi.bonet@somenergia.coop>

juan-pe committed:
[WIP] 🚧 energy_communities_cooperator: invited to member work in progress

implemented tests to check transition from invited to member memberships

Co-authored-by: Xavi Bonet  <xavi.bonet@somenergia.coop>

juan-pe committed:
[WIP] 🚧 energy_communities_cooperator: invited to member work in progress

implemented tests to check transition from invited to member memberships

Co-authored-by: Xavi Bonet  <xavi.bonet@somenergia.coop>

XavierBonetViura committed:
[IMP] ✨ Change compute _compute_membership_type() and _compute_coop_candidate()

AlvaroGRuiz committed:
[FIX] 🐛 Fix tests test__subscription_request__invited2member

AlvaroGRuiz committed:
[FIX] 🐛 Fix function assign_cooperator_register_number. Fix not invited set calculation.

Co-authored-by: Juan Pedro Sánchez <juanpe.sanchez@somenergia.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Fix function set_cooperator_effective for this case cooperative_membership.member and cooperative_membership.effective_invited

AlvaroGRuiz committed:
[FIX] 🐛 Fix field compute cooperator_type of model cooperative.membership

Co-authored-by: Juan Pedro Sánchez <juanpe.sanchez@somenergia.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Add test_post_migration_invited test

Co-authored-by: Juan Pedro Sánchez <juanpe.sanchez@somenergia.coop>

AlvaroGRuiz committed:
[I18N] 🌐 Generate news po for energy communitites cooperative

Co-authored-by: Juan Pedro Sánchez <juanpe.sanchez@somenergia.coop>

AlvaroGRuiz committed:
[FIX] 🐛 Fix test post-migration energy_communities_service_invoicing and energy_communities_cooperator modules

Co-authored-by: Juan Pedro Sánchez <juanpe.sanchez@somenergia.coop>

XavierBonetViura committed:
[I18N] 🌐 energy-communities-cooperator: Added translation using Weblate (Catalan)

XavierBonetViura committed:
[I18N] 🌐 energy-communities-cooperator: Added translation using Weblate (Basque)

XavierBonetViura committed:
[I18N] 🌐 energy-communities-cooperator: Added translation using Weblate (Spanish)

juan-pe committed:
[FIX] 🐛 energy_communities_service_invoicing: migration of invited products

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>

juan-pe committed:
[FIX] 🐛 energy_communities_cooperator: fixes in migration script

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>

juan-pe committed:
[FIX] 🐛 energy_communities_coopeator: fix migrations details for invited members

AlvaroGRuiz committed:
[FIX] 🐛 Fix invited product now is created without is_contract

AlvaroGRuiz committed:
[IMP] ✨ Upload energy_communitites_cooperatory

AlvaroGRuiz committed:
[FIX] 🐛 Fix migrations invited

AlvaroGRuiz committed:
[IMP] ✨ Change filters invited on res.partner view

AlvaroGRuiz committed:
[FIX] 🐛 Fix search is_cooperator.

juan-pe committed:
[REL] 🏗️ energy_communities_service_invoicing: bump to 16.0.0.7.4

[REL] 🏗️ energy_communities: bump to 16.0.0.7.13

[REL] 🏗️ energy_communities_api: bump to 16.0.0.4.4

[REL] 🏗️ energy_communities_cooperator: bump to 16.0.0.7.0

[REL] 🏗️ energy_communities_crm: bump to 16.0.0.3.5

[REL] 🏗️ energy_communities_service_invoicing: bump to 16.0.0.7.4

[REL] 🏗️ energy_selfconsumption: bump to 16.0.0.6.0

[DEL] 🔥 clean end whitespaces in readme

juan-pe committed:
[I18N] 🌐 fix format and translation in Request email to Release Capital for CE template

Co-authored-by: Alvaro Garcia  <alvaro.garcia@somit.coop>

AlvaroGRuiz committed:
[IMP] ✨ New field conf_cadastral_reference_readonly on selfconsumption model. And the field project_header_description now is html field and translate.

AlvaroGRuiz committed:
[IMP] ✨ Generate new .po for selfconsumption module

alvarosomit committed:
[I18N] 🌐 energy-selfconsumption: Added translation using Weblate (Spanish)

alvarosomit committed:
[I18N] 🌐 energy-selfconsumption: Added translation using Weblate (Catalan)

alvarosomit committed:
[I18N] 🌐 energy-selfconsumption: Added translation using Weblate (Basque)

alvarosomit committed:
[I18N] 🌐 energy-project: Added translation using Weblate (Spanish)

alvarosomit committed:
[I18N] 🌐 energy-project: Added translation using Weblate (Catalan)

alvarosomit committed:
[I18N] 🌐 energy-project: Added translation using Weblate (Basque)

alvarosomit committed:
[I18N] 🌐 energy-selfconsumption: Added translation using Weblate (Basque)

alvarosomit committed:
[I18N] 🌐 energy-selfconsumption: Added translation using Weblate (Spanish)

alvarosomit committed:
[I18N] 🌐 energy-selfconsumption: Added translation using Weblate (Catalan)

AlvaroGRuiz committed:
[IMP] ✨ field conf_cadastral_reference_readonly is true by default

AlvaroGRuiz committed:
[REF] ♻️  Bump version energy_project v16.0.0.2.6 and energy_selfconsumption v16.0.0.6.1

juan-pe reviewed (approved):
ueeee!!!! we do it! amizing job!! 🎉 🎉 🎉 💃🏽 💃🏽 🍺 🍺 🍺 🍺
